File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int n = triangle.size() - 1;
    for (int i = n; i > 0; i--) {
        for (int j = 0; j < i; j++) {
            triangle[i-1][j] += max(triangle[i][j], triangle[i][j+1]);
        }
    }
    return triangle[0][0];
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    
    dp = [row[:] for row in triangle]
    for i in range(len(dp) - 1):
        for j in range(i+1):
            dp[i+1][j] = max(dp[i+1][j], dp[i][j] + triangle[i+1][j])
            dp[i+1][j+1] = max(dp[i+1][j+1], dp[i][j] + triangle[i+1][j+1])
    return max(dp[-1])